util/src/test/java/io/camunda/zeebe/util/startup/StartupProcessTest.java

```diff
@@ -0,0 +1,382 @@
+import static org.mockito.Mockito.inOrder;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+import static org.mockito.Mockito.when;
+
+import io.camunda.zeebe.util.sched.ConcurrencyControl;
+import io.camunda.zeebe.util.sched.TestConcurrencyControl;
+import io.camunda.zeebe.util.sched.future.ActorFuture;
+import io.camunda.zeebe.util.sched.future.TestActorFuture;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CountDownLatch;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+@SuppressWarnings({"rawtypes", "unchecked"})
+class StartupProcessTest {
+  private static final Object STARTUP_CONTEXT =
+      new Object() {
+        @Override
+        public String toString() {
+          return "startupContext";
+        }
+      };
+
+  private static final Object SHUTDOWN_CONTEXT =
+      new Object() {
+        @Override
+        public String toString() {
+          return "shutdownContext";
+        }
+      };
+
+  private static final ConcurrencyControl TEST_CONCURRENCY_CONTROL = new TestConcurrencyControl();
+
+  static final class InvocationCountingStartupStep implements StartupStep<Object> {
+
+    private int startupInvocationCounter = 0;
+    private int shutdownInvocationCounter = 0;
+
+    int getStartupInvocationCounter() {
+      return startupInvocationCounter;
+    }
+
+    int getShutdownInvocationCounter() {
+      return shutdownInvocationCounter;
+    }
+
+    @Override
+    public String getName() {
+      return "InvocationCountingStartupStep";
+    }
+
+    @Override
+    public ActorFuture<Object> startup(final Object o) {
+      startupInvocationCounter++;
+      return completedFuture(o);
+    }
+
+    @Override
+    public ActorFuture<Object> shutdown(final Object o) {
+      shutdownInvocationCounter++;
+      return completedFuture(o);
+    }
+  }
+
+  static final class WaitingStartupStep implements StartupStep<Object> {
+
+    private final CountDownLatch startupCountdownLatch;
+    private final boolean completeWithException;
+
+    WaitingStartupStep(
+        final CountDownLatch startupCountdownLatch, final boolean completeWithException) {
+      this.startupCountdownLatch = startupCountdownLatch;
+      this.completeWithException = completeWithException;
+    }
+
+    @Override
+    public String getName() {
+      return "WaitingStartupStep";
+    }
+
+    @Override
+    public ActorFuture<Object> startup(final Object o) {
+      final var startupFuture = new TestActorFuture<>();
+      final var startupThread =
+          new Thread(
+              () -> {
+                try {
+                  startupCountdownLatch.await();
+                } catch (final InterruptedException e) {
+                  e.printStackTrace();
```

Comment:
:upside_down_face: Can we use a logger here? I know it's a test, but we should still try to remain consistent :smile: